auctions/english-auction/src/main.sw

```diff
@@ -0,0 +1,265 @@
+    /// Stores the auction information based on auction ID.
+    /// Map(auction id => auction)
+    auctions: StorageMap<u64, Option<Auction>> = StorageMap {}, 
+    
+    // TODO: Move deposits into the Auction struct when StorageMaps are
+    //       supported inside structs
+    /// Stores the deposits made based on the user and auction ID.
+    /// Map((user, auction id) => deposit)
+    deposits: StorageMap<(Identity, u64), Option<Asset>> = StorageMap {},
+    /// The total number of auctions that have ever been created.
+    total_auctions: u64 = 0,
+}
+
+impl EnglishAuction for Contract {
+    #[storage(read)]
+    fn auction_info(auction_id: u64) -> Option<Auction> {
+        storage.auctions.get(auction_id)
+    }
+
+    #[storage(read, write)]
+    fn bid(auction_id: u64, bid_asset: Asset) {
+        let auction: Option<Auction> = storage.auctions.get(auction_id);
+        require(auction.is_some(), InputError::AuctionDoesNotExist);
+
+        let mut auction = auction.unwrap();
+        let sender = msg_sender().unwrap();
+        require(sender != auction.seller, UserError::BidderIsSeller);
+        require(auction.state == State::Open && height() <= auction.end_block, AccessError::AuctionIsNotOpen);
+        require(bid_asset == auction.bid_asset, InputError::IncorrectAssetProvided);
+
+        // Combine the user's previous deposits and the current bid for the
+        // total deposits to the auction the user has made
+        let sender_deposit: Option<Asset> = storage.deposits.get((sender, auction_id));
+        let total_bid = match sender_deposit {
+            Option::Some(Asset) => {
+                bid_asset + sender_deposit.unwrap()
+            },
+            Option::None(Asset) => {
+                bid_asset
+            }
+        };
+
+        match total_bid {
+            Asset::NFTAsset(nft_asset) => {
+                transfer_nft(nft_asset, sender, Identity::ContractId(contract_id()));
+                // TODO: Remove this once StorageVec is supported in structs
+                auction.state = State::Closed;
+            },
+            Asset::TokenAsset(token_asset) => {
+                require(bid_asset.amount() == msg_amount(), InputError::IncorrectAmountProvided);
+                require(bid_asset.contract_id() == msg_asset_id(), InputError::IncorrectAssetProvided);
+                // Ensure this bid is greater than initial bid and the total deposits are greater 
+                // than the current winnning bid
+                // TODO: Move this outside the match statement once StorageVec is supported in structs
+                require(token_asset.amount >= auction.initial_price, InputError::InitialPriceNotMet);
+                require(token_asset.amount > auction.bid_asset.amount(), InputError::IncorrectAmountProvided);
+            }
+        }
+
+        // Check if reserve has been met if there is one set
+        let reserve: Option<u64> = auction.reserve_price;
+        match reserve {
```

Comment:
`if auction.reserve_price.is_some()` would make this much cleaner than the match